### DISCUSSION TITLE
LazyExtras vs config

### DISCUSSION THREAD

**Discussion by Jomik** [LazyVim/LazyVim#1674] [General]
Hey! What are your thoughts on using `LazyExtras` to manage our extras? 
I assume that this introduces the need to version control `lazyvim.json` alongside the config.
Would you recommend all extras be managed through this? Or is it just an option for some to get started more easily, but then you should "probably" switch to config managed?
I tend to keep `lang` extras in a language plugin like [here](https://github.com/Jomik/dotfiles/blob/e71bdf109896ccb12a336c3212199c4326fe56b6/.config/nvim/lua/plugins/lang/go.lua).

**Comment by folke** [comment]
That't totally up to you :)

I use `LazyExtras` now to manage all extras since it's super easy and quick to enable/disable.

You can also use both approaches. Keep some extras in your config, while enabling others in the ui.

**Reply by Jomik** [reply]
I suppose those I do not add "more" onto, could be managed in the UI.

Do you exclusively use `LazyExtras`? Your dotfiles repo is probably not up to date 😁 

Thanks again as always, some cool things happening here 😄

**Reply by folke** [reply]
I do :) Just pushed the changes to my dots

**Reply by folke** [reply]
Another advantage of letting LazyVim manage extras is priorities. Some extras need to be loaded before others to work properly.

Like edgy/aerial, or any of the dap/test extras for example.

You can manually ensure the correct order (and will be notified when not correct), but when using `LazyExtras`, order will always be correct.

**Reply by appelgriebsch** [reply]
@folke is there a way to add "custom" extension to the list shown in `LazyExtras`? And one minor caveat: in the lazyvim.json that is created there is an entry pointing to the users full path to the lazyvim folder, which may invoke some security challenges....

**Reply by folke** [reply]
Yeah, was thinking about that. Planning to implement that soon.

And totally right on the json file! It's also not iempotent and not formatted. Will work on a fix. Thanks for letting me know

**Reply by appelgriebsch** [reply]
otherwise big kudos to the 10 release. updated my `Nv` configuration w/o much hassle and with the switch to `dashboard` and `aerial` it feels even faster and slicker now

**Reply by fredrikaverpil** [reply]
Somewhat related: #1677 😄  but I've personally also switched over onto `LazyExtras` instead and stores the `lazyvim.json` in my dotfiles repo, so having formatting/idempotency to it would be awesome.

The 10 release is really something. Really stellar work, Folke!

**Reply by appelgriebsch** [reply]
as soon as the `LazyExtra` offers the proposed extension capability I likely remove my custom config scheme and move to LazyExtra for it

**Reply by folke** [reply]
I just pushed formatting/idempotency for the lazyvim.json file. Will make a release later today after I fixed some other things :)

**Reply by igorgue** [reply]
@folke how do I make my own extras in my own config? It's a great idea because I have to check-in extras but users of my config might not want that.

**Reply by folke** [reply]
That's also what @appelgriebsch was suggesting.
Will work on something later tonight. https://github.com/LazyVim/LazyVim/issues/1681

**Reply by appelgriebsch** [reply]
you can just follow the same mechanism that is used in `lazyvim` itself... e.g. https://github.com/appelgriebsch/Nv/blob/main/lua/plugins/extras/test/coverage.lua and then import it as usual https://github.com/appelgriebsch/Nv/blob/main/lua/config/lazy.lua#L75

**Reply by folke** [reply]
@appelgriebsch changes are now live. Your custom extras should be picked up automatically and shown in `LazyExtras`.

I will make a new release tomorrow. To test it now, add `version = false` here: https://github.com/appelgriebsch/Nv/blob/2056e458abf0175dd1f43fcb2d5e3d68cd7de5a9/lua/config/lazy.lua#L16

**Reply by appelgriebsch** [reply]
@folke works like a charm -> https://github.com/appelgriebsch/Nv/pull/9.... still figuring out how to customize / support themes (light/dark) before merging, but already running it from the feature branch \o/

**Reply by folke** [reply]
Great! One thing to be aware of is that your custom extras will run before your `plugins`.

**Reply by salvaft** [reply]
Hello there! I am not a power user but if i understood correctly we can register our own extras! right?
I have an use case for that (I tweaked lualine conf inside copilot extra and up to now i am forced to create config files for copilot and lua line instead of using the built in extra).

EDIT: Looking at @appelgriebsch config i just had to create a lua file under plugin/extras and it will be listed! I think this might have popped up in the update notifications but hard to capture. Would be nice to update the docs!

Thanks and happy coding

**Reply by appelgriebsch** [reply]
@folke I moved most of my "extensions" to `LazyExtra` right now (see PR from above). This includes choice for themes (onedarkpro, github at the moment) quite fine. 

Have two small questions:
1. if there is an Extra that is a substitute for another one (like the theme above, or like the aerial and symbol-outline extension from base), in which case only one should be active... is there a way to "deselect" the alternative variant from the LazyExtra when selecting the other one? otherwise we may end up with having both in the LazyExtra settings and side-effects may occur....
2. I added an option to toggle between light / dark mode in my themes and would like to "store" the last setting the user has chosen. can I use the same lazyvim.json you use for the LazyExtras settings somehow to persists this option? if so: how?

Otherwise this new setup with LazyExtras is just f---ing awesome.... never had so much fun configuring an editor / IDE :)

**Reply by appelgriebsch** [reply]
@folke I moved most of my "extensions" to `LazyExtra` right now (see PR from above). This includes choice for themes (onedarkpro, github at the moment) quite fine. 

Have two small questions:
1. if there is an Extra that is a substitute for another one (like the theme above, or like the aerial and symbol-outline extension from base), in which case only one should be active... is there a way to "deselect" the alternative variant from the LazyExtra when selecting the other one? otherwise we may end up with having both in the LazyExtra settings and side-effects may occur....
2. I added an option to toggle between light / dark mode in my themes and would like to "store" the last setting the user has chosen. can I use the same lazyvim.json you use for the LazyExtras settings somehow to persists this option? if so: how?

Otherwise this new setup with LazyExtras is just f---ing awesome.... never had so much fun configuring an editor / IDE :)

**Reply by folke** [reply]
For 1. you mean to automatically disable an extra when another one is enabled? Need to think about this, but for now users just need to not select both?

2. for now, it's better not to use that, since I may rewrite, or remove changes to that file at some point.
But you can just use a vim global with all uppercase letters for this. Global variables all uppercase are stored in the shada file automatically, so perfect for your use-case

**Reply by folke** [reply]
And thanks :)

**Reply by appelgriebsch** [reply]
Worked. Released 1.0.0 of [Nv](https://github.com/appelgriebsch/Nv) just now \o/

**Reply by folke** [reply]
Nice! Congrats! :)

**Comment by qwertzui11** [comment]
I added all essentials extras, like testing and debugging, to the lua config 
```lua
require("lazy").setup({
  spec = {
    -- add "core" LazyVim plugins
    { "LazyVim/LazyVim", import = "lazyvim.plugins" },
    { import = "lazyvim.plugins.extras.dap.core" },
    { import = "lazyvim.plugins.extras.test.core" },
    { import = "lazyvim.plugins.extras.util.project" },

    { import = "plugins" },
  },
-- [...]
})
```

However all language extras like 
```json
    "lazyvim.plugins.extras.lang.clangd",
    "lazyvim.plugins.extras.lang.cmake",
    "lazyvim.plugins.extras.lang.docker",
    "[...]"
```
are managed by the LazyExtras.

This way I'm able to handle the `lazyvim.json` as a local file for each machine I work on.

For example:
At work I use mainly `C++`, so I enable the extras `clangd` and `cmake` using `LazyExtras` on my work-machine.
Back home I mostly develop `Rust`, so I enable the extra `rust`.

This keeps the plugin list shorter and side-effects smaller. Eg mason is not maintaining an unneeded LSP.

So, I don't version the `lazyvim.json` and added it to my ignore list.

Any thoughts on my approach :question: :smiley: 

@folke Great work, btw! :heart_eyes_cat:

**Reply by igorgue** [reply]
I did the same and ignored it, great to have many machines on the same config but different purposes, for example the extra I have for swift I'd enable it only on my mac, same goes for the omnisharp one for windowze.

I wonder if the file could be ignored as well @folke checked it in in his dotfiles.

**Comment by 0x6a68** [comment]
I also really enjoy configuring extras via `lazyvim.json`. 

I would also be in favor of beeing able to configure `LazyExtras` via a `lazyvim.local.json` to have extras enable per project and not only per install/machine. Wdyt? 

p.s. I hope this is not a necrobump 😨

**Comment by arkochan** [comment]
Is there any way to quickly generate the list of extras that are enabled now? Something similar to the `pip freeze` command, I mean. 
In that case we can just update the `require("lazy").setup`'s `spec` table accordingly when we see fit and use version control as usual. 
I know the list isn't super difficult to generate but to maintain with version control it'd need to be generated again and again. So, this can be a way. 👌

**Reply by Jomik** [reply]
You can just read the `extras` array from the config, using something like `cat lazyvim.json | jq '.extras'` for a JSON array or `cat lazyvim.json | jq '.extras[]' -r` to get a line separated list.
Doesn't really do a freeze - that would be part of the lockfile, and hard to know what is from an extra and what is not.
